### PR TITLE
Code Quality: Improve string comparer performance

### PR DESCRIPTION
**Pull request by viruseg**
New comparator is slightly faster than [DllImport(“api-ms-win-core-string-l1-1-0.dll”)].CompareStringEx. And supports Arm. Previously, a string comparer was used for Arm without proper support for natural comparison.

| Method                                 | Mean     | Error     | StdDev    | Allocated |
|--------------------------------------- |---------:|----------:|----------:|----------:|
| Test_api-ms-win-core-string-l1-1-0.dll | 1.185 ms | 0.0095 ms | 0.0084 ms |      89 B |
| Test_NaturalComparer                   | 1.072 ms | 0.0073 ms | 0.0065 ms |      89 B |

### CODE REVIEW

**Comment by yaira2** [comment]
@XTorLukas fyi

**Review by Imran-imtiaz48 — COMMENTED** [review]
Strengths:
1.	Code Structure:
o	The overall structure is clean and follows good practices. Separating the logic into different Compare overloads makes the functionality clear and modular.
o	Usage of sealed for NaturalComparer is excellent as it enforces immutability and ensures that no one can derive further classes from it.
2.	Natural Sorting:
o	The natural sorting logic is well-implemented. Handling numeric values and trimming leading zeros demonstrates attention to detail for real-world comparisons.
3.	Extensibility:
o	The inclusion of multiple IComparer implementations for objects, strings, and memory spans makes the class highly reusable in diverse scenarios.
4.	Performance Optimization:
o	Use of ReadOnlySpan<char> for comparisons is a nice touch, ensuring memory efficiency and reducing allocations.
Suggestions for Improvement:
1.	Error Handling:
o	It might be useful to add some error handling or documentation on expected behavior for edge cases, such as comparing incompatible object types that are not strings or IComparable.
2.	Code Duplication:
o	The logic in Compare(string? x, string? y) and Compare(ReadOnlyMemory<char> x, ReadOnlyMemory<char> y) is nearly identical. Consider consolidating them to avoid duplication, perhaps by internally calling the Span overload.
3.	Test Coverage:
o	While the logic is robust, it’s important to ensure this is accompanied by comprehensive test cases to verify behavior across a variety of inputs: 
	Strings with leading/trailing spaces.
	Numeric strings of varying lengths.
	Mixed strings with numbers and letters (e.g., abc123 vs abc23).
4.	Default Behavior:
o	Consider documenting or adding a safeguard for cases where x or y are not string or IComparable. Right now, it defaults to StringComparer.FromComparison, which might lead to unexpected results.
5.	Inline Documentation:
o	While the code is clear, inline comments explaining some of the more complex parts (e.g., CompareNumValues logic for leading zeros) would improve readability for future maintainers.
6.	Static Default:
o	Instead of instantiating NaturalComparer multiple times, you could provide a static readonly instance for common comparisons (e.g., Ordinal).
Code Example for Consolidation:
Here’s how you could consolidate the Compare methods:
public int Compare(ReadOnlyMemory<char> x, ReadOnlyMemory<char> y) 
    => Compare(x.Span, y.Span, stringComparison);

public int Compare(string? x, string? y) 
    => Compare(x.AsSpan(), y.AsSpan(), stringComparison);

Summary
This is a solid implementation of a natural comparer with attention to edge cases like leading zeros and numeric sorting. A few tweaks around code reuse, documentation, and test coverage would make it even better. Great job overall.

File: src/Files.App/Helpers/NaturalStringComparer.cs [hunk 1]
```diff
@@ -7,32 +7,110 @@ public sealed class NaturalStringComparer
 	{
 		public static IComparer<object> GetForProcessor()
 		{
-			return Win32Helper.IsRunningOnArm ? new StringComparerArm64() : new StringComparerDefault();
+			return new NaturalComparer(StringComparison.CurrentCulture);
 		}
-
-		private sealed class StringComparerArm64 : IComparer<object>
+
```
Comment (catuhana) [inline src/Files.App/Helpers/NaturalStringComparer.cs:12]: Whitespace left here.

File: src/Files.App/Helpers/NaturalStringComparer.cs [hunk 2]
```diff
@@ -7,32 +7,110 @@ public sealed class NaturalStringComparer
 	{
 		public static IComparer<object> GetForProcessor()
 		{
-			return Win32Helper.IsRunningOnArm ? new StringComparerArm64() : new StringComparerDefault();
+			return new NaturalComparer(StringComparison.CurrentCulture);
 		}
-
-		private sealed class StringComparerArm64 : IComparer<object>
+		
+		private sealed class NaturalComparer : IComparer<object?>, IComparer<string?>, IComparer<ReadOnlyMemory<char>>
```
Comment (0x5bfa) [inline src/Files.App/Helpers/NaturalStringComparer.cs:13]: Can you add xml comment on this class so others can understand what this is doing under the hood.

File: src/Files.App/Helpers/NaturalStringComparer.cs [hunk 3]
```diff
@@ -7,32 +7,110 @@ public sealed class NaturalStringComparer
...
+
+		        if (num1AsSpan.Length > num2AsSpan.Length) return 1;
+
+		        var compareResult = num1AsSpan.CompareTo(num2AsSpan, StringComparison.Ordinal);
+
+		        if (compareResult != 0) return Math.Sign(compareResult);
+
+		        if (numValue2.Length == numValue1.Length) return compareResult;
+
+		        return numValue2.Length < numValue1.Length ? -1 : 1; // "033" < "33" === true
```
Comment (0x5bfa) [inline src/Files.App/Helpers/NaturalStringComparer.cs:112]:
```suggestion
		        return numValue2.Length < numValue1.Length ? -1 : 1; // "033" < "33" == true
```

**Review by 0x5bfa — CHANGES_REQUESTED** [review]

**Comment by 0x5bfa** [comment]
FYI, this isn't yet published and documented but C# runtime just superseded `CompareOptions.NumericOrdering` 2 weeks ago.

**Comment by 0x5bfa** [comment]
FYI, C# runtime just superseded this with `CompareOptions.NumericOrdering` 2 weeks ago (not documented yet).

**Comment by 0x5bfa** [comment]
C# runtime implemented SortOptions.NumericOrdering just two weeks ago. We will use it once we upgrade to NET 9.x, instead.

**Review by yaira2 — APPROVED** [review]
LGTM
